describe_log_dirs_request.go

```diff
@@ -21,34 +21,68 @@ type DescribeLogDirsRequestTopic struct {
 }
 
 func (r *DescribeLogDirsRequest) encode(pe packetEncoder) error {
-	length := len(r.DescribeTopics)
-	if length == 0 {
-		// In order to query all topics we must send null
-		length = -1
-	}
-
-	if err := pe.putArrayLength(length); err != nil {
-		return err
-	}
+	isFlexible := r.Version >= 2
 
-	for _, d := range r.DescribeTopics {
-		if err := pe.putString(d.Topic); err != nil {
+	if isFlexible {
+		if len(r.DescribeTopics) == 0 {
+			pe.putUVarint(0)
+		} else {
+			pe.putCompactArrayLength(len(r.DescribeTopics))
+		}
```

Comment:
This code works out to be the same as below, except with `putCompactArrayLength` instead of `putArrayLength`:
```suggestion
		length := len(r.DescribeTopics)
		if length == 0 {
			length = -1
		}
		pe.putCompactArrayLength(length)
```

In fact, the two code paths are now mostly the same and we can just use:
```go
	length := len(r.DescribeTopics)
	if length == 0 {
		// In order to query all topics we must send null
		length = -1
	}
	if isFlexible {
		pe.putCompactArrayLength(length)
	} else {
		if err := pe.putArrayLength(length); err != nil {
			return err
		}
	}
```

Reply:
I was trying to be consistent with the slightly similar code in `offset_fetch_request.go` handling of `partitions` but I agree the suggested code is much nicer (and functionally the same). (I prefer the strict `nil` check `offset_fetch_request.go` uses for `partitions` to the `len` check used here.)